### PR TITLE
Dec 756 map config to api

### PR DESCRIPTION
**Pull request by jonhartzler**
Refactor to make the metadata configuration configurable per collection.

Basic What:

1.  Changed all out going references to configuration to ask the collection what its config is.  
2.  Made saving and updating work again.  By changing name to be a true metadata field.   
3.  Changed a lot of tests to reference doubles rather than the items.yml config.
4.  Changed the app to use the items.yml to setup a default config value. 

Still to do 
Remove public interface to metadata in item.
Make Description act like name.
Move configuration from its own database table to a part of collection. 
Move user_defined_id out of metadata.

### CODE REVIEW

File: app/controllers/v1/collections_controller.rb [hunk 1]
```diff
@@ -55,7 +55,9 @@ def preview_mode
     end
 
     def metadata_configuration
-      @configuration = Metadata::Configuration.item_configuration
+      collection = CollectionQuery.new.any_find(params[:collection_id])
+      @configuration = CollectionConfigurationQuery.new(collection).find
+      @configuration = Metadata::Configuration.new(@configuration)
```
Comment (rfox2) [inline app/controllers/v1/collections_controller.rb]: what happens during the successive mutation?

File: app/models/item.rb [hunk 2]
```diff
@@ -34,8 +36,8 @@ class Item < ActiveRecord::Base
   has_attached_file :uploaded_image,
                     restricted_characters: /[&$+,\/:;=?@<>\[\]{}\|\\^~%#]/
 
-  validates :name, :collection, :unique_id, :user_defined_id, presence: true
-  validates :date_created, :date_modified, :date_published, date: true
+  validates :collection, :unique_id, :user_defined_id, presence: true
+  # validates :date_created, :date_modified, :date_published, date: true
```
Comment (rfox2) [inline app/models/item.rb]: can this be deleted?

File: app/models/item.rb [hunk 3]
```diff
@@ -43,6 +45,14 @@ class Item < ActiveRecord::Base
 
   enum image_status: { image_invalid: 0, image_processing: 1, image_ready: 2 }
```
Comment (rfox2) [inline app/models/item.rb]: unrelated to pull request, but are we still using "image_invalid" as a descriptor?

File: app/controllers/v1/collections_controller.rb (hunk 1 above)
Comment (jonhartzler) [inline app/controllers/v1/collections_controller.rb]: This is a non dry usage that I have corrected in the next Pull Request.

File: app/models/item.rb (hunk 2 above)
Comment (jonhartzler) [inline app/models/item.rb]: YES !!

File: app/models/item.rb (hunk 3 above)
Comment (jonhartzler) [inline app/models/item.rb]: no it is no_image  that is in the pull request i merged in the morning.

File: app/models/metadata/retrieval.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def metadata_value(field)
+      value = get(field)
+      config = configuration.field(field)
+
+      return nil if value.nil? || config.nil?
+
+      type = config.type
+      value = ensure_value_is_array(value)
+
+      if type.to_sym == :string
```
Comment (rfox2) [inline app/models/metadata/retrieval.rb]: could this conditional sequence become a private method?

File: app/models/waggle.rb [hunk 5]
```diff
@@ -24,6 +24,19 @@ def self.commit
   end
 
   def self.search(**args)
+    set_configuration(::Metadata::Configuration.new(CollectionConfigurationQuery.new(args[:collection]).find))
```
Comment (rfox2) [inline app/models/waggle.rb]: noob question: why does Metadata::Configuration need to be namespaced this way?

File: app/services/create_items.rb [hunk 6]
```diff
@@ -28,7 +28,7 @@ def create_or_update!(collection_id:, find_by:, items_hash:, counts:, errors:)
     ActiveRecord::Base.transaction do
       items_hash.each.with_index do |item_props, index|
         rewrite_errors = []
-        item_props = yield(item_props, rewrite_errors) if block_given?
+        item_props = yield(item_props, rewrite_errors).symbolize_keys if block_given?
```
Comment (rfox2) [inline app/services/create_items.rb]: when you get a chance could you explain how this expression works?

File: app/models/waggle.rb (hunk 5 above)
Comment (jonhartzler) [inline app/models/waggle.rb]: I did it to follow what was already started with Jaron's work.   He created a metadata section.   I have been trying to use that more.  Also, Metadata indicates it is in a directory.  So to create a directory I have to do this namespacing.

File: app/models/metadata/retrieval.rb (hunk 4 above)
Comment (jonhartzler) [inline app/models/metadata/retrieval.rb]: I thought about that I will do it.

File: app/services/metadata_input_cleaner.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+class MetadataInputCleaner
```
Comment (rfox2) [inline app/services/metadata_input_cleaner.rb]: not to be picky - this name seems a little confusion. this service class is essentially decomposing an array value into the item metadata values using a hash.  how about MetadataInputDecomposition - it's a mouthful, i know

File: spec/models/waggle/adapters/solr/search/result_spec.rb [hunk 8]
```diff
@@ -109,8 +115,7 @@
 
       it "is the configured facet fields, excluding their relevant filter" do
         expect(subject).to eq([
-          "{!ex=creator_facet}creator_facet",
-          "{!ex=language_facet}language_facet"
+          "{!ex=creator_facet}creator_facet"
```
Comment (rfox2) [inline spec/models/waggle/adapters/solr/search/result_spec.rb]: I think rubocop just wants the left and right brackets to be on their own line with the values indented inside of the array brackets

**Comment by rfox2** [comment]
I just had one remaining question about the naming of MetadataInputCleaner, but that's not a show stopper. All other questions are answered! :-)

Once the hound issues are looked at I'd say :+1:

File: app/models/item.rb [hunk 9]
```diff
@@ -34,8 +36,7 @@ class Item < ActiveRecord::Base
   has_attached_file :uploaded_image,
                     restricted_characters: /[&$+,\/:;=?@<>\[\]{}\|\\^~%#]/
 
-  validates :name, :collection, :unique_id, :user_defined_id, presence: true
```
Comment (jgondron) [inline app/models/item.rb]: Is something else now handling validation of these fields? Even if we're moving them to metadata, if the expectation was that the app was validating their presence, then we should continue to do so.

Comment (jonhartzler) [inline app/models/item.rb]: Yes.  The metadata configuration manages this.  All things metadata need to be managed in metadata not duplicated into places that no longer have or manage that field.   Name is no longer considered a part of item.  The only reason that it is mapped in  is for some of the admin views that display name and description (think the item list).   It would be fixed by using the api completely.

File: app/services/create_items.rb (hunk 6 above)
Comment (jgondron) [inline app/services/create_items.rb]: Wondering if the injected block shouldn't handle symbolizing the keys. The purpose of these injected blocks is to transform the item hash into something we can directly pass to an item save, so it seems reasonable to expect it to symbolize the keys.

File: spec/controllers/v1/collections_controller_spec.rb [hunk 10]
```diff
@@ -210,10 +237,5 @@
       expect(SaveCollection).to receive(:call).with(collection, site_objects: site_objects_translated).and_return(true)
       subject
     end
-
-    it "checks the editor permissions" do
```
Comment (jgondron) [inline spec/controllers/v1/collections_controller_spec.rb]: Why was this removed?

File: spec/decorators/v1/item_json_decorator_spec.rb [hunk 11]
```diff
@@ -26,7 +26,7 @@
   end
 
   describe "#at_id" do
-    let(:item) { double(Item, unique_id: "adsf") }
+    let(:item) { double(Item, unique_id: "adsf", name: "name") }
```
Comment (jgondron) [inline spec/decorators/v1/item_json_decorator_spec.rb]: I'm still a little concerned with using double instead of instance_double. When we remove the store_accessors from Item these specs will still pass, which doesn't give us very good regression testing.

File: spec/validators/date_validator_spec.rb [hunk 12]
```diff
@@ -1,5 +1,8 @@
 require "rails_helper"
 
+# commented out because it is not in use.  Not removed because I think I want to bring it
+# back in the next refactor
+=begin
```
Comment (jgondron) [inline spec/validators/date_validator_spec.rb]: Do not comment about usage of block comments.

File: spec/controllers/v1/collections_controller_spec.rb (hunk 10 above)
Comment (jonhartzler) [inline spec/controllers/v1/collections_controller_spec.rb]: Because metadata configuration is not an editor only action.

File: spec/decorators/v1/item_json_decorator_spec.rb (hunk 11 above)
Comment (jonhartzler) [inline spec/decorators/v1/item_json_decorator_spec.rb]: updated

File: spec/validators/date_validator_spec.rb (hunk 12 above)
Comment (jonhartzler) [inline spec/validators/date_validator_spec.rb]: I just wanted it to be clear why I did not just remove this file.   Because the whole file is commented out.
